libraries/botbuilder-ai/botbuilder/ai/qna/qnamaker.py

```diff
@@ -197,132 +244,12 @@ def _validate_options(self, options: QnAMakerOptions):
-        headers = self._get_headers()
-
-        # Convert miliseconds to seconds (as other BotBuilder SDKs accept timeout value in miliseconds)
-        # aiohttp.ClientSession units are in seconds
-        timeout = ClientTimeout(total=options.timeout / 1000)
-
-        response = await self._req_client.post(
-            url, data=serialized_content, headers=headers, timeout=timeout
-        )
-
-        result = await self._format_qna_result(response, options)
-
-        return result
-
-    async def _emit_trace_info(
-        self, turn_context: TurnContext, result: [QueryResult], options: QnAMakerOptions
-    ):
-        trace_info = QnAMakerTraceInfo(
-            message=turn_context.activity,
-            query_results=result,
-            knowledge_base_id=self._endpoint.knowledge_base_id,
-            score_threshold=options.score_threshold,
-            top=options.top,
-            strict_filters=options.strict_filters,
-        )
-
-        trace_activity = Activity(
-            label=QNAMAKER_TRACE_LABEL,
-            name=QNAMAKER_TRACE_NAME,
-            type="trace",
-            value=trace_info,
-            value_type=QNAMAKER_TRACE_TYPE,
-        )
-
-        await turn_context.send_activity(trace_activity)
-
-    async def _format_qna_result(
-        self, result, options: QnAMakerOptions
-    ) -> [QueryResult]:
-        json_res = result
-        if isinstance(result, ClientResponse):
-            json_res = await result.json()
-
-        answers_within_threshold = [
-            {**answer, "score": answer["score"] / 100}
-            for answer in json_res["answers"]
-            if answer["score"] / 100 > options.score_threshold
-        ]
-        sorted_answers = sorted(
-            answers_within_threshold, key=lambda ans: ans["score"], reverse=True
-        )
-
-        # The old version of the protocol returns the id in a field called qnaId
-        # The following translates this old structure to the new
-        if self._is_legacy_protocol:
-            for answer in answers_within_threshold:
-                answer["id"] = answer.pop("qnaId", None)
-
-        answers_as_query_results = list(
-            map(lambda answer: QueryResult(**answer), sorted_answers)
-        )
-
-        return answers_as_query_results
-
-    def _get_headers(self):
-        headers = {
-            "Content-Type": "application/json",
-            "User-Agent": self.get_user_agent(),
-        }
-
-        if self._is_legacy_protocol:
-            headers["Ocp-Apim-Subscription-Key"] = self._endpoint.endpoint_key
-        else:
-            headers["Authorization"] = f"EndpointKey {self._endpoint.endpoint_key}"
-
-        return headers
+    def _has_matched_answer_in_kb(self, query_results: [QueryResult]) -> bool:
+        if query_results:
+            if len(query_results) is not None:
+                ans = query_results[0].answer
+                if ans != "No good match found in KB.":
```

Comment:
Instead of comparing with  _"No good match found in KB."_ string, which can be changed by user, we can have a check if   `query_results[0].id != -1`